### PR TITLE
Add `stage` argument to tell checking functions what type of check is being performed

### LINKED ISSUE #567: Exercise checking should include an argument indicating which type of check is being perfomed
If we add a ~`type`~ `stage` argument to the function signature of the exercise checker, `learnr` can communicate which checking step is being performed.

Practically speaking this would allow `gradethis` to relax the "always return a grade" constraint for code checking, which would allow grading to do a first-pass code check before full evaluation.

The impact would be small for exercise checking functions that include `...`. Alternatively, the checking step can be inferred from the values of particular arguments as described in [Exercises - Custom Checking](https://rstudio.github.io/learnr/exercises.html#Custom_Checking), but it'd be cleaner to simply tell the checker what step is being performed.

cc @schloerke

### PR DESCRIPTION
**Pull request by rossellhayes**
Calls to `try_checker()` within `evaluate_exercise()` gain the argument `stage` – one of `"code_check"`, `"error_check"`, or `"check"` – which is passed to the checking function to indicate which kind of check is being performed.

This simplifies the implementation of custom checking functions, which no longer need to infer the stage from other arguments. For example, the example function in the `learnr` docs becomes:

```r
custom_checker <- function(label, user_code, check_code, envir_result, evaluate_result, last_value, stage, ...) {
  # this is a code check
  if (stage == "code_check") {
    if (is_bad_code(user_code, check_code)) {
      return(list(message = "I wasn't expecting that code", correct = FALSE))
    }
    return(list(message = "Nice code!", correct = TRUE))
  }
  # this is a fully evaluated chunk check
  if (is_bad_result(last_value, check_code)) {
    return(list(message = "I wasn't expecting that result", correct = FALSE))
  }
  list(message = "Great job!", correct = TRUE, location = "append")
}
```

For backwards compatibility, if the function signature of the checking function does not include `stage` or `...`, `stage` is not passed.

Closes #567.

### CODE REVIEW

**Review by gadenbuie — APPROVED** [review]
Looks great, thank you! I think the changes in the docs do a good job showcasing the improvement!

**Comment by gadenbuie** [comment]
Oh and let's bump the dev version for this one

File: R/exercise.R [hunk 1]
```diff
@@ -501,17 +505,23 @@ try_checker <- function(
...
   # Throw better error messaging if the checker function signature is ill-defined
   missing_args <- setdiff(names(args), checker_args)
   if (length(missing_args) && !"..." %in% checker_args) {
-    msg <- sprintf(
-      "Either add ... or the following arguments to the '%s' function: '%s'",
-      name, paste(missing_args, collapse = "', '")
-    )
-    message(msg)
-    rlang::return_from(rlang::caller_env(), exercise_result_error(msg))
+    # Don't throw an error if the only missing argument is `stage`
```
[the remark this answers is not in the document]
Comment (rossellhayes) [reply R/exercise.R]: How does 5b36cf1 look?

**Comment by rossellhayes** [comment]
@gadenbuie Dev version bumped! Should this go in the `NEWS` as a "new feature" or a "minor new feature"?

**Pull request merged**